From f50b36e2a10aae367719a385a9b421265728c9b4 Mon Sep 17 00:00:00 2001
From: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
Date: Wed, 3 Apr 2019 13:01:45 -0700
Subject: [PATCH 0596/1851] v4l: xilinx: scd: Merge the main and V4L2 IRQ
 handlers

commit b223d84fe847fd49e16a04d5ee8695d0a08925cf from
https://github.com/Xilinx/linux-xlnx.git

Both the main SCD driver and the V4L2 support code register an IRQ
handler, for the same IRQ. This is unnecessary, we can just call the
V4L2 IRQ handling function from the main IRQ handler.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
Signed-off-by: Satish Kumar Nagireddy <satish.nagireddy.nagireddy@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 .../xilinx/xilinx-scenechange-channel.c       | 48 ++++---------------
 .../platform/xilinx/xilinx-scenechange.c      |  5 ++
 .../platform/xilinx/xilinx-scenechange.h      |  3 +-
 3 files changed, 14 insertions(+), 42 deletions(-)

diff --git a/drivers/media/platform/xilinx/xilinx-scenechange-channel.c b/drivers/media/platform/xilinx/xilinx-scenechange-channel.c
index 9f2100eff0a6..232716bfa062 100644
--- a/drivers/media/platform/xilinx/xilinx-scenechange-channel.c
+++ b/drivers/media/platform/xilinx/xilinx-scenechange-channel.c
@@ -343,53 +343,19 @@ static void xscd_event_notify(struct xscd_chan *chan)
 	v4l2_subdev_notify_event(&chan->subdev, &chan->event);
 }
 
-static irqreturn_t xscd_chan_irq_handler(int irq, void *data)
+void xscd_chan_irq_handler(struct xscd_chan *chan)
 {
-	struct xscd_chan *chan = (struct xscd_chan *)data;
-	struct xscd_shared_data *shared_data;
-
-	shared_data = &chan->xscd->shared_data;
+	struct xscd_shared_data *shared_data = &chan->xscd->shared_data;
 	spin_lock(&chan->dmachan.lock);
 
 	if ((shared_data->memory_based && chan->dmachan.valid_interrupt) ||
 	    !shared_data->memory_based) {
 		spin_unlock(&chan->dmachan.lock);
 		xscd_event_notify(chan);
-		return IRQ_HANDLED;
+		return;
 	}
 
 	spin_unlock(&chan->dmachan.lock);
-	return IRQ_NONE;
-}
-
-static int xscd_chan_parse_of(struct xscd_chan *chan)
-{
-	struct xscd_shared_data *shared_data;
-	int err;
-
-	shared_data = &chan->xscd->shared_data;
-	shared_data->dma_chan_list[chan->id] = &chan->dmachan;
-	chan->iomem = shared_data->iomem;
-
-	chan->irq = irq_of_parse_and_map(chan->xscd->dev->of_node, 0);
-	if (!chan->irq) {
-		dev_err(chan->xscd->dev, "No valid irq found\n");
-		return -EINVAL;
-	}
-
-	err = devm_request_irq(chan->xscd->dev, chan->irq,
-			       xscd_chan_irq_handler, IRQF_SHARED,
-			       dev_name(chan->xscd->dev), chan);
-	if (err) {
-		dev_err(chan->xscd->dev, "unable to request IRQ %d\n",
-			chan->irq);
-		return err;
-	}
-
-	chan->dmachan.iomem = shared_data->iomem;
-	chan->dmachan.id = chan->id;
-
-	return 0;
 }
 
 /**
@@ -417,9 +383,11 @@ int xscd_chan_init(struct xscd_device *xscd, unsigned int chan_id,
 	mutex_init(&chan->lock);
 	chan->xscd = xscd;
 	chan->id = chan_id;
-	ret = xscd_chan_parse_of(chan);
-	if (ret < 0)
-		return ret;
+	chan->iomem = chan->xscd->iomem;
+	chan->dmachan.id = chan->id;
+	chan->dmachan.iomem = chan->xscd->iomem;
+
+	xscd->shared_data.dma_chan_list[chan->id] = &chan->dmachan;
 
 	/* Initialize V4L2 subdevice and media entity */
 	subdev = &chan->subdev;
diff --git a/drivers/media/platform/xilinx/xilinx-scenechange.c b/drivers/media/platform/xilinx/xilinx-scenechange.c
index 3d0f2b71279e..bcf216e5acb8 100644
--- a/drivers/media/platform/xilinx/xilinx-scenechange.c
+++ b/drivers/media/platform/xilinx/xilinx-scenechange.c
@@ -16,6 +16,7 @@
 static irqreturn_t xscd_irq_handler(int irq, void *data)
 {
 	struct xscd_device *xscd = (struct xscd_device *)data;
+	unsigned int i;
 	u32 status;
 
 	status = xscd_read(xscd->iomem, XSCD_ISR_OFFSET);
@@ -23,6 +24,10 @@ static irqreturn_t xscd_irq_handler(int irq, void *data)
 		return IRQ_NONE;
 
 	xscd_write(xscd->iomem, XSCD_ISR_OFFSET, XSCD_IE_AP_DONE);
+
+	for (i = 0; i < xscd->numstreams; ++i)
+		xscd_chan_irq_handler(xscd->chans[i]);
+
 	return IRQ_HANDLED;
 }
 
diff --git a/drivers/media/platform/xilinx/xilinx-scenechange.h b/drivers/media/platform/xilinx/xilinx-scenechange.h
index ded93b21f042..6e00c3e9acc8 100644
--- a/drivers/media/platform/xilinx/xilinx-scenechange.h
+++ b/drivers/media/platform/xilinx/xilinx-scenechange.h
@@ -145,7 +145,6 @@ static inline struct xscd_dma_chan *to_xscd_dma_chan(struct dma_chan *chan)
 
 /**
  * struct xscd_chan - Video Stream structure
- * @irq: device IRQ
  * @id: scene change channel ID
  * @iomem: device I/O register space remapped to kernel virtual memory
  * @xscd: SCD device
@@ -157,7 +156,6 @@ static inline struct xscd_dma_chan *to_xscd_dma_chan(struct dma_chan *chan)
  * @lock: lock to protect active stream count variable
  */
 struct xscd_chan {
-	int irq;
 	int id;
 	void __iomem *iomem;
 	struct xscd_device *xscd;
@@ -263,6 +261,7 @@ void xscd_dma_chan_enable(struct xscd_dma_chan *chan, int chan_en);
 void xscd_dma_reset(struct xscd_dma_chan *chan);
 void xscd_dma_halt(struct xscd_dma_chan *chan);
 
+void xscd_chan_irq_handler(struct xscd_chan *chan);
 int xscd_chan_init(struct xscd_device *xscd, unsigned int chan_id,
 		   struct device_node *node);
 #endif
-- 
2.31.1

